allennlp/training/util.py

```diff
@@ -378,17 +379,31 @@ def rescale_gradients(model: Model, grad_norm: Optional[float] = None) -> Option
     return None
 
 
-def get_metrics(
-    model: Model, total_loss: float, num_batches: int, reset: bool = False
-) -> Dict[str, float]:
+def get_metrics(model: Model,
+                total_loss: float,
+                num_batches: int,
+                reset: bool = False,
+                world_size: int = 1,
+                rank: int = 0) -> Dict[str, float]:
     """
     Gets the metrics but sets ``"loss"`` to
     the total loss divided by the ``num_batches`` so that
     the ``"loss"`` metric is "average loss per batch".
     """
     metrics = model.get_metrics(reset=reset)
     metrics["loss"] = float(total_loss / num_batches) if num_batches > 0 else 0.0
-    return metrics
+
+    if world_size > 1:
```

Comment:
I think this needs to be before the bit above where we replace the `metrics["loss"]` with a float, as it won't have been all_reduced (and then subsequently won't be all reduced correctly, as we've lost the reference to the tensor)

Reply:
Sorry, I don't quite get your point here. `total_loss` is a float and I'm not sure what reference we'd be losing. Essentially, `total_loss / num_batches` give the effective loss in a worker and we need to log the average loss across workers. Aggregation happens for all the metrics in the `dict` and hence the effective loss is being added there before the `for` loop. 
Could you elaborate a bit here?

Reply:
Ah I see - the loss has already been all-reduced for the backward pass so here we only care about the value.

Reply:
Strictly speaking, it is the gradients that have been all-reduced during the backward pass.